src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.cs

```diff
@@ -92,20 +92,16 @@ private LambdaCompiler(AnalyzedTree tree, LambdaExpression lambda, MethodBuilder
             var scope = tree.Scopes[lambda];
             var hasClosureArgument = scope.NeedsClosure;
 
-            Type[] paramTypes = GetParameterTypes(lambda);
-            if (hasClosureArgument)
-            {
-                paramTypes = paramTypes.AddFirst(typeof(Closure));
-            }
+            Type[] paramTypes = GetParameterTypes(lambda, hasClosureArgument ? typeof(Closure) : null);
 
             method.SetReturnType(lambda.ReturnType);
             method.SetParameters(paramTypes);
-            var paramNames = lambda.Parameters.Map(p => p.Name);
```

Comment:
No need to allocate an array for this, only to enumerate it below. As mentioned earlier, together with another PR, this reduces the need for `Map`.